### DISCUSSION TITLE
Welcome to postfix-policy-server Discussions!

### DISCUSSION THREAD

**Discussion by wneessen** [wneessen/postfix-policy-server#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by samrocks1011** [comment]
Hi Wneessen,

I wanted to know if we can send the bounce to the sender in case the policy matched. Currenty the mail get rejected without sending any notfication to the sender.

Regards,
Sandeep Patel

**Reply by wneessen** [reply]
Hi Sandeep,

this depends a bit on the configuration of your postfix and the sending server. But generally a "REJECT" action should trigger a bounce mail on the sending end. If you are using a "DISCARD" or s. th. similar, the mail will be discard without letting the sender know.

REJECT action: https://github.com/wneessen/postfix-policy-server/blob/main/pps.go#L42

Also check the postfix access(5) man page for details: https://www.postfix.org/access.5.html

Hope this answers your question. Otherwise please provide some more details or even better an example code snippet.

**Reply by samrocks1011** [reply]
HI Wneesen,

The REJECT action is not generating the bounce.  I have set the below change in Postfix for policy to work.

smtpd_recipient_restrictions = check_policy_service inet:127.0.0.1:10005

If we see the Postfix log the mail get rejected and disconnect the connection and print the message that I have specfied in main.go without generating any bounce.  

Postfix Log
Feb 12 18:45:36 osticket postfix/smtpd[519954]: NOQUEUE: reject: RCPT from tools.wormly.com[172.104.20.135]: 554 5.7.1 <example@gmail.com>: Recipient address rejected: customers@wormly.com is blocked!; from=<customers@wormly.com> to=<example@gmail.com> proto=ESMTP helo=<www.wormly.com>
Feb 12 18:45:37 osticket postfix/smtpd[519954]: lost connection after RCPT from tools.wormly.com[172.104.20.135]


Regards,
Sandeep Patel



main.go 
--------------
import (
        "context"
        "encoding/json"
        "fmt"
        "github.com/wneessen/postfix-policy-server"
        "log"
)

// Hi is an empty struct to work as the Handler interface
type Hi struct{}

// Handle is the test handler for the test server as required by the Handler interface
func (h Hi) Handle(ps *pps.PolicySet) pps.PostfixResp {
        log.Println("received new policy set...")
        jps, err := json.Marshal(ps)
        if err != nil {
                log.Printf("failed to marshal policy set data: %s", err)
                return pps.RespWarn
        }
        fmt.Println(" SENDER NAME:",ps.Sender);
        fmt.Println(string(jps))
        if ps.Sender =="customers@wormly.com" {
                return pps.TextResponseOpt(pps.RespReject, "customers@wormly.com is blocked!")
        }else{
                if ps.Recipient  =="example@gmail.com"{
                        return pps.TextResponseOpt(pps.RespReject, "Recepient not allowed")
                }else{
                        return pps.TextResponseOpt(pps.RespOk, "The email is okay to be delived!")
                }
                //return pps.TextResponseOpt(pps.RespOk, "The email is okay to be delived!")
        }
        return pps.TextResponseOpt(pps.RespInfo, "this might be a cool mail!")
}

// main starts the server
func main() {
        s := pps.New()
        ctx, cancel := context.WithCancel(context.Background())
        defer cancel()
        h := Hi{}
        log.Println("Starting policy echo server...")
        if err := s.Run(ctx, h); err != nil {
                log.Fatalf("could not run server: %s", err)
        }
}
--------------------------------

**Reply by wneessen** [reply]
I've just tried to reproduce with your code example and this should generate a bounce:

```
$ echo -e "EHLO test.de
MAIL FROM: <customers@wormly.com>
RCPT TO: <wneessen>
QUIT" | nc localhost 25
220 arch-vm.localdomain ESMTP Postfix
250-arch-vm.localdomain
250-PIPELINING
250-SIZE 10240000
250-VRFY
250-ETRN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-DSN
250-SMTPUTF8
250 CHUNKING
250 2.1.0 Ok
554 5.7.1 <wneessen>: Recipient address rejected: customers@wormly.com is blocked!
221 2.0.0 Bye
```

As you can see in the 2nd to last line, it returned with a 5.x.x error, which should tell the sending server to return a bounce notification to the sender. If this does not happen for you, this is either due to some configuration on your postfix server (not generating a 5.x.x error for a REJECT) or due to some configuration on the sending side, that will not send out a bounce notification in case a mail could not be delivered.

If you want to force a notification to the sender (which I would not advice to do), you could have your policy server generate a mail in case a mail was reject.

Maybe something like this:
![image](https://user-images.githubusercontent.com/542696/153725186-8dd7e1a5-5034-4dfe-9211-ac264d5943d4.png)

**Reply by samrocks1011** [reply]
Hi,

Thanks for the help. Bounce was generated successfully some issues in postfix routing. I reinstalled the postfix and the issue got resolved.  

I also wanted to create a Content filter rule but seems check_policy_service do not provide Mail Header and Body in the attribute. I thing I have to move to a postfix content filter(http://www.postfix.org/FILTER_README.html) for the same.

Regards,
Sandeep Patel
